scipy/optimize/_simulated_annealing.py

```diff
@@ -0,0 +1,557 @@
+        if self.state_improved:
+            # Global energy has improved, let's see if LS improved further
+            e, x = self.obj_fun_wrapper.local_search(self.state.xbest)
+            if e < self.state.ebest:
+                self.not_improved_idx = 0
+                self.state.ebest = e
+                self.state.xbest = np.copy(x)
+                self.state.current_energy = e
+                self.state.current_location = np.copy(x)
+                return
+        # Check probability of a need to perform a LS even if no improvment
+        # (Dual annealing principle)
+        do_ls = False
+        if self.K < 90 * len(self.state.current_location):
+            pls = np.exp(self.K * (self.state.ebest - self.state.current_energy
+                                   ) / self.temperature_step)
+            if pls >= self._rand_state.random_sample():
+                do_ls = True
+        # Global energy not improved, let's see what LS gives
+        # on the best Markov chain location
+        if self.not_improved_idx >= self.not_improved_max_idx:
+            do_ls = True
+        if do_ls:
+            e, x = self.obj_fun_wrapper.local_search(self.xmin)
+            self.xmin = np.copy(x)
+            self.emin = e
+            self.not_improved_idx = 0
+            self.not_improved_max_idx = self.state.current_location.size
+            if e < self.state.ebest:
+                self.state.ebest = self.emin
+                self.state.xbest = np.copy(self.xmin)
+                self.state.current_energy = e
+                self.state.current_location = np.copy(x)
+
+
+class ObjectiveFunWrapper(object):
+    """
+    Class used to wrap around the objective function in order to apply local
+    search.
+    Default local minimizer is SciPy minimizer L-BFGS-B
+    """
+    
+    MAXITER_RATIO = 6
+    MAXITER_MIN = 100
+    MAXITER_MAX = 1000
+
+    def __init__(self, bounds, func, **kwargs):
+        self.func = func
+        self.nb_fun_call = 0
+        self.kwargs = kwargs
+        self.minimizer = minimize
+        self.fun_args = ()
+        bounds_list = list(zip(*bounds))
+        self.lower = np.array(bounds_list[0])
+        self.upper = np.array(bounds_list[1])
+        if 'args' in self.kwargs:
+            self.fun_args = self.kwargs.get('args')
+        
+    def set_default_minimizer(self):
+        # By default, use SciPy minimize with 'L-BFGS-B' method
+        n = len(self.lower)
+        ls_max_iter = min(max(n * self.MAXITER_RATIO, self.MAXITER_MIN),
+                          self.MAXITER_MAX)
+        self.kwargs['method'] = 'L-BFGS-B'
+        self.kwargs['options'] = {
+            'maxiter': ls_max_iter,
+        }
+        self.kwargs['bounds'] = list(zip(self.lower, self.upper))
+
+    def local_search(self, x):
+        x_tmp = np.copy(x)
+        fun_temp = self.func(x, *self.fun_args)
+        mres = self.minimizer(self.func, x, *self.fun_args, **self.kwargs)
+        self.nb_fun_call += mres.nfev 
```

Comment:
what happens if the minimizer doesn't support bounds.

Reply:
Solved by addressing the comment above.
The minimizer can use unbounded space but the implementation checks that the results is in bounds otherwise, it is rejected.